osbuild/util/pathfd.py

```diff
@@ -0,0 +1,716 @@
+    require either access right).
+    With path-descriptors you can refer to paths in a stable manner without
+    requiring read nor write access to the path.
+
+  * ...
+
+The path-descriptors are functionally superior to normal string-based path
+objects, but not necessarily the better option in all situations. Their
+flexibility and stable behavior is a big strength, but their biggest weakness
+is the overhead involved in all the syscalls to manage them. They are based on
+`O_PATH` file-descriptor, rather than simple strings, and thus require syscalls
+for every single operation. In many situations these additional syscalls
+amortize easily, but that has to be evaluated for every situation separately.
+
+Please be aware that path-descriptors are not suitable for access management!
+Access to a path-descriptor does **NOT** restrict the access you gain via it.
+You can always use relative operations on the path-descriptor to access its
+parents, children, and any other part of the file-system reachable from that
+path.
+You can, however, combine path-descriptors with other technology to achieve
+access restrictions. This implementation allows for this, but considers the
+discussion of this topic beyond the scope of this module.
+"""
+
+
+import errno as _errno
+import os as _os
+import stat as _stat
+
+
+__all__ = [
+    "PathFd",
+]
+
+
+class PathFd:
+    """Path Descriptor
+
+    A path-descriptor object is a representation of a file-system path. Every
+    instance represents exactly one path. Internally, every instance owns its
+    backing `O_PATH` file-descriptor, which pins the capability of referring to
+    the path it represents.
+
+    For a discussion of when to use path-descriptors over simple paths, see the
+    `O_PATH` functionality provided by linux. This path-descriptor simply
+    provides easy access to `O_PATH` file-descriptors.
+
+    Since path-descriptors manage resources, they do provide the standard
+    resource management operations. This means, `close()` releases all internal
+    resources and turns the instance into a no-op stub. Further operations on
+    the stub will trigger assertions. This follows what the python standard
+    library provides, and thus integrates well with context-managers (a default
+    context manager is provided with the implementation, but
+    `contextlib.closing()` would also work.
+    Note that there is no requirement to call `close()`. The only reason to do
+    so is to keep resource management under caller-control if they so wish. If
+    the GC cleanup is enough for you, then you can avoid calling `close()`. But
+    if you want to avoid having stale resources pile up until the next GC
+    round, then you better close your objects when no longer needed.
+    The internal operations of this module always clean up all their temporary
+    resources explicitly to avoid this problem.
+    """
+
+    _fd = None
```

Comment:
If you don't do that, the `__del__` destructor might be called without `_fd` being defined. This can happen if you create an instance, but pass an argument-list that does not match the `__init__` arguments.

Granted, this only happens during development, but it does avoid a very confusing error in `__del__` in that case.